src/refract/JSONSchemaVisitor.cc

```diff
@@ -7,646 +7,664 @@
+    }
 
-        if (fixed || fixedType) {
-            addMember("additionalProperties", IElement::Create(false));
-        }
+    if (fixed || fixedType) {
+        addMember("additionalProperties", make_primitive(false));
     }
+}
 
-    void JSONSchemaVisitor::anyOf(
-        std::map<std::string, std::vector<IElement*> >& types, std::vector<std::string>& typesOrder)
-    {
-        ArrayElement* a = new ArrayElement;
+void JSONSchemaVisitor::anyOf(
+    std::map<std::string, std::vector<const IElement*> >& types, std::vector<std::string>& typesOrder)
+{
+    if (typesOrder.empty()) {
+        addMember("anyOf", make_empty<ArrayElement>());
+
+    } else {
+        auto a = make_element<ArrayElement>();
 
         for (auto const& item : typesOrder) {
 
-            const std::vector<IElement*>& items = types[item];
+            const auto& items = types[item];
 
-            IElement* elm = items.front();
+            const IElement* elm = items.front();
             JSONSchemaVisitor v(pDefs);
             Visit(v, *elm);
 
             if (TypeQueryVisitor::as<EnumElement>(elm)) {
                 v.addMember("enum", elm->clone());
             } else if (!TypeQueryVisitor::as<ObjectElement>(elm)) {
-                ArrayElement::ValueType enmVals;
-                CloneMembers(&enmVals, &items);
+                data::array_t enmVals;
+
+                CloneMembers(enmVals, items);
 
                 if (!enmVals.empty()) {
-                    ArrayElement* enm = new ArrayElement;
-                    enm->set(enmVals);
-                    v.addMember("enum", enm);
+                    v.addMember("enum", make_element<ArrayElement>(std::move(enmVals)));
                 }
             }
 
-            a->push_back(v.getOwnership());
+            a->get().push_back(v.getOwnership());
         }
-        addMember("anyOf", a);
+        addMember("anyOf", std::move(a));
     }
+}
 
-    void JSONSchemaVisitor::operator()(const ArrayElement& e)
-    {
-        const ArrayElement::ValueType* val = GetValue<ArrayElement>(e);
+void JSONSchemaVisitor::operator()(const ArrayElement& e)
+{
+    const ArrayElement* v = GetValue<ArrayElement>{}(e);
 
-        if (!val || val->empty()) {
-            return;
-        }
+    if (!v)
+        return;
+    if (v->empty())
+        return;
 
-        JSONSchemaVisitor renderer(pDefs);
-        setSchemaType("array");
+    const auto& val = v->get();
+    if (val.empty())
+        return;
 
-        if (IsTypeAttribute(e, "fixed")) {
-            fixed = true;
-        }
+    JSONSchemaVisitor renderer(pDefs);
+    setSchemaType("array");
 
-        if (IsTypeAttribute(e, "fixedType")) {
-            fixedType = true;
-        }
+    if (IsTypeAttribute(e, "fixed")) {
+        fixed = true;
+    }
 
-        if (fixed || fixedType) {
-            ArrayElement::ValueType av;
-            bool allEmpty = allItemsEmpty(val);
+    if (IsTypeAttribute(e, "fixedType")) {
```

Comment:
We probably should rename this to `HasTypeAttribute()`
I know it is not "your" code, but while review I found it :)